Allow TestCluster to stop an individual Server.

mrtracy opened the pull request:
Adds "StopServer" method to TestCluster, allowing individual TestServers
in a TestCluster to be stopped independently.

This commit primarily changes the "stopper" behavior in TestCluster. Previously,
one stopper was shared between all TestServers. Now, each server is given its
own stopper; the cluster creates a Worker task that waits until the cluster's
Stopper is Stop()ed, and then closes the individual stoppers of the servers.

Added a test to verify that a single server is stopped.

Commit additionally fixes a bug where open GRPC connections to a TestServer
would still function after it had been stopped.

<!-- Reviewable:start -->

---

This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/8075)

<!-- Reviewable:end -->

Commit:
Allow TestCluster to stop an individual Server.

Adds "StopServer" method to TestCluster, allowing individual TestServers
in a TestCluster to be stopped independently.

This commit primarily changes the "stopper" behavior in TestCluster. Previously,
one stopper was shared between all TestServers. Now, each server is given its
own stopper; the cluster creates a Worker task that waits until the cluster's
Stopper is Stop()ed, and then closes the individual stoppers of the servers.

Added a test to verify that a single server is stopped.

tamird commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> but let @andreimatei have a look

---

Reviewed 5 of 5 files at r1.
Review status: all files reviewed at latest revision, 7 unresolved discussions, some commit checks failed.

---

*[server/server.go, line 320 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8075#-KNjf6-krwXagHNiKQ0e:-KNjf6-lyinZOMHAuuTb:1379992607) ([raw file](https://github.com/cockroachdb/cockroach/blob/1f5e3e93138ce165ab07b78106bb5367ffca3622/server/server.go#L320)):*
> ```Go
> 	s.stopper.RunWorker(func() {
> 		<-s.stopper.ShouldQuiesce()
> 		if err := ln.Close(); err != nil {
> ```

can you put the grpc server stopping in this worker?

---

*[testutils/testcluster/testcluster.go, line 49 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8075#-KNjfxMQZprrPNnb7V7Z:-KNjfxMQZprrPNnb7V7_:177628577) ([raw file](https://github.com/cockroachdb/cockroach/blob/1f5e3e93138ce165ab07b78106bb5367ffca3622/testutils/testcluster/testcluster.go#L49)):*
> ```Go
> 	stopper *stop.Stopper
> 	mu      struct {
> 		syncutil.Mutex
> ```

why is this synchronization needed?

---

*[testutils/testcluster/testcluster.go, line 89 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8075#-KNjfRmhDsez9IDZtO5W:-KNjfRmhDsez9IDZtO5X:-957174461) ([raw file](https://github.com/cockroachdb/cockroach/blob/1f5e3e93138ce165ab07b78106bb5367ffca3622/testutils/testcluster/testcluster.go#L89)):*
> ```Go
> 
> 	}
> 	tc.mu.serverStoppers = []*stop.Stopper{}
> ```

uh, why?

---

*[testutils/testcluster/testcluster.go, line 156 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8075#-KNjfhx6h--jGELx490C:-KNjfhx6h--jGELx490D:-377210643) ([raw file](https://github.com/cockroachdb/cockroach/blob/1f5e3e93138ce165ab07b78106bb5367ffca3622/testutils/testcluster/testcluster.go#L156)):*
> ```Go
> 	// Create a worker that will close the individual server stoppers when the
> 	// cluster stopper is stopped.
> 	tc.stopper.RunWorker(func() {
> ```

you can use `tc.stopper.AddCloser(stop.CloserFunc(tc.stopServers))` instead of a worker

---

*[testutils/testcluster/testcluster_test.go, line 194 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8075#-KNjgG-qoNsb8rTt0wWt:-KNjgG-qoNsb8rTt0wWu:328843225) ([raw file](https://github.com/cockroachdb/cockroach/blob/1f5e3e93138ce165ab07b78106bb5367ffca3622/testutils/testcluster/testcluster_test.go#L194)):*
> ```Go
> 	var response serverpb.HealthResponse
> 
> 	httpClient, err := server1.GetHTTPClient()
> ```

call this httpClient1 for symmetry with `server1` and ditto for `url`

---

*[testutils/testcluster/testcluster_test.go, line 221 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8075#-KNjg8fmXmK-RgfMDXmF:-KNjg8fmXmK-RgfMDXmG:-228762479) ([raw file](https://github.com/cockroachdb/cockroach/blob/1f5e3e93138ce165ab07b78106bb5367ffca3622/testutils/testcluster/testcluster_test.go#L221)):*
> ```Go
> 
> 	// Verify HTTP and GRPC requests to server now fail.
> 	if err := util.GetJSON(httpClient, url, &response); err == nil {
> ```

check for a specific error and log it if it's not the expected one.

---

*[testutils/testcluster/testcluster_test.go, line 225 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8075#-KNjgA5_l1tycxF8QDWI:-KNjgA5_l1tycxF8QDWJ:-469765787) ([raw file](https://github.com/cockroachdb/cockroach/blob/1f5e3e93138ce165ab07b78106bb5367ffca3622/testutils/testcluster/testcluster_test.go#L225)):*
> ```Go
> 	}
> 
> 	if _, err := adminClient.Health(ctx, &serverpb.HealthRequest{}); err == nil {
> ```

ditto

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/8075#-:-KNjgMtTvs8HkskB2IoV:1126920686)*
<!-- Sent from Reviewable.io -->

RaduBerinde commented:
LGTM (modulo Tamir's comments)

---

Reviewed 5 of 5 files at r1.
Review status: all files reviewed at latest revision, 8 unresolved discussions, some commit checks failed.

---

*[server/server.go, line 369 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8075#-KNmVnG-uW-MtIRW0XrF:-KNmVnG-uW-MtIRW0XrG:1299981941) ([raw file](https://github.com/cockroachdb/cockroach/blob/1f5e3e93138ce165ab07b78106bb5367ffca3622/server/server.go#L369)):*
> ```Go
> 		netutil.FatalIfUnexpected(s.grpc.Serve(anyL))
> 
> 		// GRPC server will stop accept new connections when the anyL listener
> ```

accepting

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/8075#-:-KNmXNyB5pT9Ln-pyM_D:-681300671)*
<!-- Sent from Reviewable.io -->

bdarnell commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

---

Reviewed 3 of 5 files at r1.
Review status: all files reviewed at latest revision, 8 unresolved discussions, some commit checks failed.

---

*[testutils/testcluster/testcluster.go, line 49 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8075#-KNjfxMQZprrPNnb7V7Z:-KNmCvQiR1xchRqeCNgW:267415648) ([raw file](https://github.com/cockroachdb/cockroach/blob/1f5e3e93138ce165ab07b78106bb5367ffca3622/testutils/testcluster/testcluster.go#L49)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> why is this synchronization needed?
</details>
`StopServer` replaces a stopper with nil. The synchronization mirrors a similar pattern in `multiTestContext`. It may not be necessary for the current use of this feature but seems like a good idea. 

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/8075#-:-KNmXtqSz3aZD8_Q5DOK:1417088721)*
<!-- Sent from Reviewable.io -->

mrtracy commented:
Review status: 2 of 4 files reviewed at latest revision, 8 unresolved discussions, some commit checks pending.

---

*[testutils/testcluster/testcluster.go, line 49 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8075#-KNjfxMQZprrPNnb7V7Z:-KO6OsPPedl89RTup5VB:-38102930) ([raw file](https://github.com/cockroachdb/cockroach/blob/1f5e3e93138ce165ab07b78106bb5367ffca3622/testutils/testcluster/testcluster.go#L49)):*
<details><summary><i>Previously, bdarnell (Ben Darnell) wrote…</i></summary>
> `StopServer` replaces a stopper with nil. The synchronization mirrors a similar pattern in `multiTestContext`. It may not be necessary for the current use of this feature but seems like a good idea. 
</details>
Just a precaution, stopping a stopper twice is a panic.

---

*[testutils/testcluster/testcluster.go, line 89 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8075#-KNjfRmhDsez9IDZtO5W:-KO6P3I9wHXD3_MviamD:2093317422) ([raw file](https://github.com/cockroachdb/cockroach/blob/1f5e3e93138ce165ab07b78106bb5367ffca3622/testutils/testcluster/testcluster.go#L89)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> uh, why?
</details>
I supposed it's unnecessary, removed.

---

*[testutils/testcluster/testcluster.go, line 156 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8075#-KNjfhx6h--jGELx490C:-KO6PZO-XqUuYyDgeYlA:-499146009) ([raw file](https://github.com/cockroachdb/cockroach/blob/1f5e3e93138ce165ab07b78106bb5367ffca3622/testutils/testcluster/testcluster.go#L156)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> you can use `tc.stopper.AddCloser(stop.CloserFunc(tc.stopServers))` instead of a worker
</details>
Done.

---

*[testutils/testcluster/testcluster_test.go, line 194 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8075#-KNjgG-qoNsb8rTt0wWt:-KO6PrqgrfqG2ZyWs-s3:-499146009) ([raw file](https://github.com/cockroachdb/cockroach/blob/1f5e3e93138ce165ab07b78106bb5367ffca3622/testutils/testcluster/testcluster_test.go#L194)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> call this httpClient1 for symmetry with `server1` and ditto for `url`
</details>
Done.

---

*[testutils/testcluster/testcluster_test.go, line 221 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8075#-KNjg8fmXmK-RgfMDXmF:-KO6fR8w8WfrtaOksGSM:-499146009) ([raw file](https://github.com/cockroachdb/cockroach/blob/1f5e3e93138ce165ab07b78106bb5367ffca3622/testutils/testcluster/testcluster_test.go#L221)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> check for a specific error and log it if it's not the expected one.
</details>
Done.

---

*[testutils/testcluster/testcluster_test.go, line 225 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8075#-KNjgA5_l1tycxF8QDWI:-KO6fRXZbGHcU4pFIDyL:-499146009) ([raw file](https://github.com/cockroachdb/cockroach/blob/1f5e3e93138ce165ab07b78106bb5367ffca3622/testutils/testcluster/testcluster_test.go#L225)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> ditto
</details>
Done.

---

*[server/server.go, line 320 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8075#-KNjf6-krwXagHNiKQ0e:-KO6OdspjuiD-4K4kL7G:-675029095) ([raw file](https://github.com/cockroachdb/cockroach/blob/1f5e3e93138ce165ab07b78106bb5367ffca3622/server/server.go#L320)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> can you put the grpc server stopping in this worker?
</details>
Looks like you managed to get GracefulStop() to work, so i've just removed it entirely.

---

*[server/server.go, line 369 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8075#-KNmVnG-uW-MtIRW0XrF:-KO6OrMpfNEsOLc4CtlP:-716907597) ([raw file](https://github.com/cockroachdb/cockroach/blob/1f5e3e93138ce165ab07b78106bb5367ffca3622/server/server.go#L369)):*
<details><summary><i>Previously, RaduBerinde wrote…</i></summary>
> accepting
</details>
Removed as this change was no longer needed.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/8075)*
<!-- Sent from Reviewable.io -->

RaduBerinde commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

---

Review status: 2 of 4 files reviewed at latest revision, 7 unresolved discussions, some commit checks pending.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/8075#-:-KO6jyioRyqANPTIDMBE:1417088721)*
<!-- Sent from Reviewable.io -->

tamird commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

---

Reviewed 3 of 3 files at r2.
Review status: all files reviewed at latest revision, 1 unresolved discussion, some commit checks pending.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/8075#-:-KO6kigdH1JM1s8hPoPO:1417088721)*
<!-- Sent from Reviewable.io -->
